### PR TITLE
Switch array impls to const generics

### PR DESCRIPTION
**Pull request by austinabell**
Necessary changes to make https://github.com/paritytech/ink/issues/752 happen

### CODE REVIEW

File: src/codec.rs [hunk 1]
```diff
@@ -688,57 +688,32 @@ impl_for_non_zero! {
...
-					}
-				}
-			}
+impl<T: Decode, const N: usize> Decode for [T; N] {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let mut r = ArrayVec::new();
+		for _ in 0..N {
+			r.push(T::decode(input)?);
+		}
+		let i = r.into_inner();
```
Comment (bkchr) [inline src/codec.rs:707]:
```suggestion
```

File: src/codec.rs [hunk 2]
```diff
@@ -688,57 +688,32 @@ impl_for_non_zero! {
...
-						Err(_) => Err("failed to get inner array from ArrayVec".into()),
-					}
-				}
-			}
+impl<T: Decode, const N: usize> Decode for [T; N] {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let mut r = ArrayVec::new();
+		for _ in 0..N {
+			r.push(T::decode(input)?);
+		}
```
Comment (bkchr) [inline src/codec.rs:706]:
```suggestion
		let mut array = ArrayVec::new();
		for _ in 0..N {
			array.push(T::decode(input)?);
		}
```

File: src/codec.rs [hunk 3]
```diff
@@ -688,57 +688,32 @@ impl_for_non_zero! {
...
+		let i = r.into_inner();
 
-			impl<T: EncodeLike<U>, U: Encode> EncodeLike<[U; $n]> for [T; $n] {}
-		)*
+		match i {
+			Ok(a) => Ok(a),
+			Err(_) => Err("failed to get inner array from ArrayVec".into()),
+		}
```
Comment (bkchr) [inline src/codec.rs:712]:
```suggestion
		Ok(array.into_inner().expect("We decode `N` elements; qed"))
```

**Review by bkchr — APPROVED** [review]

File: src/codec.rs (hunk 3 above)
Comment (austinabell) [reply src/codec.rs]: Can't apply this one because the error type (the `ArrayVec`) doesn't implement debug. Lmk if you want another form like `if let`

Comment (bkchr) [reply src/codec.rs]: Okay fine :)

**Pull request merged**